### ISSUE TITLE
different live_update steps for the same image

### ISSUE THREAD

**Issue by nicks** [tilt-dev/tilt#3385]
In the #tilt channel, Robby Pocase asks:

> Docker-compose use case - is there a way to have two of the same image running and have different live_reload options?  I have two containers which include equivalent source code. One hosts a running application and one that just does the equivalent of jest --watch  for running the application tests.

> Due to a lib dependency, something is breaking --watch on sync, so I'd like to do restart_container() (or whatever the current recommended equivalent is) on the test container, but not the application container.

I've heard other versions of this. e.g., you have one big "mega-image" with all your services, and you invoke it with different commands in the k8s yaml

The best way to do this right now is to use dependent images: one base image, and then have images with different names that depend on it -- they don't add any layers, they just add the live-update steps.

I don't have a good idea for this now, but wanted to log it in case there are other use-cases I'm not thinking of

**Comment by mattfenwick** [comment]
I ran into a similar scenario: I have a deployment and a job that both use the same image, but need different entrypoints.

**Comment by milas** [comment]
Similar use case came up with a user today where an image is used 2x - once for a "setup job" and once for the actual service. Things got unintentionally broken due to modifying the `entrypoint` arg on `docker_build` and having it affect both.

**Comment by nicks** [comment]
Here are some random thoughts about this because @lizzthabet was asking about it. This is kind of a brain dump:

1) Suppose you have different live-update rules, list-a and list-b, for the same image name. Without loss of generality, we can assume that there is some file A that is being live-updated by list-a, but not by list-b. Therefore, the image with live-update rules list-b are not keeping the image up-to-date properly.

MAYBE this is OK. Maybe, at the time you wrote the rules, you have some confidence that the image will never read that file. But what if that changes? What if someone, at a later time, writes some code that assumes file A is synced, and is confused by why the live-update isn't working?

If Tilt allows people to do this, it feels like you're setting their team up for a maintenance headache, because it's hard to reason about why list-a and list-b are different.

2) Increasingly, in the container image community, having one big image with multiple entrypoints is considered an anti-pattern, cf http://jpetazzo.github.io/2021/11/30/docker-build-container-images-antipatterns/

3) Increasingly, the container image community is investing in tooling to make it easier to re-tag images with different names. These tools are the most low-effort workaround for the above problem, because they let you give a single image multiple names for use in different contexts. Meanwhile, in the background, you can The reason we're all doing this is because it gives you a good transition path from one image w/ multiple entrypoints. See here: https://github.com/ImJasonH/ImJasonH/tree/main/articles/moving-and-building-images

4) Similarly to point (3), container optimization tools assume a single entrypoint, as in https://github.com/docker-slim/docker-slim

5) Tilt uses the image name to determine where live-updates attach. More complicated attachment schemes would necessarily require more complicated configuration. For example `k8s_custom_deploy` allows you to attach live updates by attaching the container name rather than the container image. Even that wouldn't cover all use-case. I suspect you'd need a patchwork of new configuration options to make this work.

TLDR - for most users, the FAR better path is to figure out how to give the images different names, which sets them up better for how the rest of the ecosystem wants to be used.

**Comment by zkutasi** [comment]
Thanks @nicks for the detailed explanations. I agree, my usage is specific that I use Tilt in my organization just for development, and the purpose is NOT to change anything production-wise, even if it is an antipattern, even if it is not best practice. Let the organization by other means strive towards these solutions. So I would not want to change things how we handle this is Docker. But luckly I was able to find a solution, so it can be worked around, thanks to the infinite flexibility of the Tiltfile :).

I agree with you, then this ticket can be closed, as the solution is clear.
